modules/phase_field/src/userobjects/PolycrystalVoronoi.C

```diff
@@ -166,6 +209,59 @@ PolycrystalVoronoi::precomputeGrainStructure()
         _centerpoints[grain](2) = _bottom_left(2) + _range(2) * 0.5;
     }
   }
+
+  // Build a KDTree that is used to speedup point search
+  buildSearchTree();
+}
+
+void
+PolycrystalVoronoi::buildSearchTree()
+{
+  if (!_use_kdtree)
+    return;
+
+  auto midplane = _bottom_left + _range / 2.0;
+  dof_id_type lgrianid = 0;
+  _grain_gtl_ids.clear();
+  _grain_gtl_ids.reserve(_centerpoints.size() * std::pow(2.0, _mesh.dimension()));
+  _new_points.clear();
+  // Use more memory. Factor is 2^dim
+  _new_points.reserve(_centerpoints.size() * std::pow(2.0, _mesh.dimension()));
+  // Domain will be extended along the periodic directions
+  // For each direction, a half domain is constructed
+  std::vector<std::vector<Real>> xyzs(3);
+  for (auto & point : _centerpoints)
+  {
+    // Cear up
+    for (unsigned int i = 0; i < 3; i++)
```

Comment:
Not a huge fan of hard-coded integers for looping.

Reply:
What is it the reliable way to check we are living a 3D space? Libmesh_dim?

Reply:
`LIBMESH_DIM` is fine here. (anything but 3 for LIBMESH_DIM will break moose anyways...)